Emby.Server.Implementations/Library/Validators/CollectionPostScanTask.cs

```diff
@@ -0,0 +1,137 @@
+using MediaBrowser.Model.Querying;
+using Jellyfin.Data.Enums;
+using Microsoft.Extensions.Logging;
+using MediaBrowser.Model.Entities;
+
+namespace Emby.Server.Implementations.Library.Validators
+{
+    /// <summary>
+    /// Class CollectionPostScanTask.
+    /// </summary>
+    public class CollectionPostScanTask : ILibraryPostScanTask
+    {
+        private readonly ILibraryManager _libraryManager;
+        private readonly ICollectionManager _collectionManager;
+        private readonly ILogger<CollectionPostScanTask> _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CollectionPostScanTask" /> class.
+        /// </summary>
+        /// <param name="libraryManager">The library manager.</param>
+        /// <param name="collectionManager">The collection manager.</param>
+        /// <param name="logger">The logger.</param>
+        public CollectionPostScanTask(
+            ILibraryManager libraryManager,
+            ICollectionManager collectionManager,
+            ILogger<CollectionPostScanTask> logger)
+        {
+            _libraryManager = libraryManager;
+            _collectionManager = collectionManager;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Runs the specified progress.
+        /// </summary>
+        /// <param name="progress">The progress.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>Task.</returns>
+        public async Task Run(IProgress<double> progress, CancellationToken cancellationToken)
+        {
+            var boxSets = _libraryManager.GetItemList(new InternalItemsQuery
+            {
+                IncludeItemTypes = new[] { nameof(BoxSet) },
+                CollapseBoxSetItems = false,
+                Recursive = true
+            });
+
+            var collectionNameMoviesMap = new Dictionary<string, HashSet<Guid>>();
+
+            foreach (var library in _libraryManager.RootFolder.Children)
+            {
+                if (!_libraryManager.GetLibraryOptions(library).AutoCollection) {
+                    continue;
+                }
+
+                var movies = _libraryManager.GetItemList(new InternalItemsQuery
+                {
+                    MediaTypes = new string[] { MediaType.Video }, 
+                    IncludeItemTypes = new[] { nameof(Movie) },
+                    IsVirtualItem = false,
+                    OrderBy = new[] { (ItemSortBy.SortName, SortOrder.Ascending) },
+                    Parent = library,
+                    Recursive = true
+                });
+
+                foreach (var m in movies)
+                {
+                    if (m is Movie movie && !string.IsNullOrEmpty(movie.CollectionName))
+                    {
+                        if (collectionNameMoviesMap.TryGetValue(movie.CollectionName, out var movieList))
+                        {
+                            if (!movieList.Contains(movie.Id))
```

Comment:
Don’t need to check if the element exists in a hash set before adding it, the add function already does the check

https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.add?view=net-6.0

Reply:
Fixed. Thanks.